teps/0025-hermekton.md

```diff
@@ -76,28 +75,41 @@ This option should not be used to run otherwise untrusted code without another l
 
 ### API Changes
 
-Tekton Pipelines will add support for a new "ExecutionMode" field on several objects.
-That type will look like:
+Tekton Pipelines will allow `TaskSpec`s to specify special "fence" steps that allow execution to transition into or out of "hermetic" execution mode:
 
-```go
-type ExecutionMode struct {
-	Hermetic bool
-}
+```yaml
+steps:
+- name: fetch-stuff
+  ...
+
+# This would direct Tekton to start the network jail.
+- name: tekton-begin-hermetic
+
+- name: build-stuff
+  ...
+
+# This would direct Tekton to disable the network jail.
+- name: tekton-end-hermetic
+
+- name: publish-stuff
+  ...
 ```
 
-This currently holds just a single bool, but could be expanded in the future.
-See [this rationale](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#primitive-types) in the k8s API style guide for why we introduce a new type.
+This change introduces the idea of special steps where a semantic is specified
+by using a particular step `name:`, **without** an `image:`.  It might make
+sense eventually to support parameterizing these semantics with `args:`, but
+the two special steps being introduced here will not support that today:
 
+* `name: tekton-begin-hermetic`: Will enable the network jail for subsequent steps.
+* `name: tekton-end-hermetic`: Will disable the network jail for subsequent steps.
```

Comment:
this is an interesting option - at the very least i think it makes sense to include the original proposal as an alternative, but also i prefer the idea of having steps explicitly marked as hermetic or not hermetic vs. what seems like dummy steps that imply that instead, e.g. something like:

```yaml
steps:
- name: fetch-stuff
  ...
- name: build-stuff
   externalAccess: hermetic
  ...
- name: publish-stuff
  ...
```

I like:
- having explicit control at a step level (and as mentioned above, potentially at the level of a task, and other levels as well)
- not overloading the meaning of a "step" to sometimes mean something which is not a step (if i understand correctly, the presence of these special steps would actually translate into specific network settings for the steps between them?)

Maybe it's worth listing the pros and cons of each approach?

Reply:
The main pro of the fence-based approach that I can see is that there is no API change required.
Since our unit of re-use if the `Task`, I would like to think we can always keep the hermetic execution flag at `TaskRun` or `Pipeline` level.

Reply:
> The main pro of the fence-based approach that I can see is that there is no API change required.

I don't think there is any particular reason we want to avoid API changes though; in fact I think introducing new functionality WITHOUT changing the API is kind of confusing (and truthfully we _are_ changing the API in that we're changing the meaning of existing fields)